Term entry swiftui accessibility (#3093)
Codecademy/docs: content/swiftui/concepts/accessibility/accessibility.md

---
Title: 'Accessibility'
Description: 'Accessibility in SwiftUI refers to the built-in tools and available techniques to create user interfaces for everyone, including individuals with diverse needs.'
Subjects:
  - 'iOS'
  - 'Mobile Development'
Tags:
  - 'SwiftUI'
  - 'iOS'
CatalogContent:
  - 'learn-swift'
  - 'paths/build-ios-apps-with-swiftui'
---

> **[Accessibility](https://www.codecademy.com/resources/docs/uiux/accessibility)** in Swift UI refers to the built-in tools and available techniques to create user interfaces for everyone, including individuals with diverse needs.

Reviewer: [proofreading] While the phrasing "everyone, including individuals with diverse needs" can be good in highlighting the importance of making Swift apps accessible, consider using the phrase below. It conveys this message clearly.  
```suggestion
**[Accessibility](https://www.codecademy.com/resources/docs/uiux/accessibility)** in Swift UI refers to the built-in tools and available techniques to create user interfaces for individuals with diverse needs.
```

    ```swift
    var body: some View {
        TextField("Username", text: $username)
            .focused($username, equals: true)
    }
    ```

    This will render a `TextField`, and when the view appears, the keyboard will automatically be shown, and the cursor will be placed in the `TextField`, ready for user input.

- Using modifiers for Dynamic Type: SwiftUI supports Dynamic Type, which means your app's text can adapt to users' chosen text size:

    ```swift
    var body: some View {
        Text("Accessibility in SwiftUI")
            .font(.body)
            .scaledFont(size: 18)
    }
    ```

    In this example above, the `Text` element is initially set with the [`.font()`](https://www.codecademy.com/resources/docs/swiftui/viewmodifier/font) modifier with `.body` value, and then the `.scaledFont()` modifier is applied to adjust the font size to `18 points`. This ensures the text remains legible even as the user changes their preferred text size settings.

> - Using semantic Views and Tags: Using semantic views and tags (e.g., `.button`, `.image`, `.heading`, and `.link`) for role-specific styling can aid the visual appearance and interactions of the app while also enhancing screen reader users' navigation experiences.

Reviewer: [Proofreading] Describing these tags as a list instead of additional information in parentheses would have developers understand that they can use these elements to make their apps more accessible. 
```suggestion
- Using semantic Views and Tags: Semantic views and tags like `.button`, `.image`, `.heading`, and `.link` for role-specific styling can aid the visual appearance and interactions of the app while also enhancing screen reader users' navigation experiences.
```

## Best Practices for Accessibility

Along with accessibility modifiers and techniques, the following best practices can help developers create apps that provide a more accessible and enjoyable user experience:

- Comprehensive Testing:
  - Regularly test your app with screen readers like VoiceOver to ensure proper labelling and conveyed information.
  - Verify keyboard navigation functionality for all interactive elements.
  - Test with various Dynamic Type settings to ensure readable text and adaptive layouts.
  -Perform contrast testing for text and graphics visibility.
- Real-World Simulations:
  - Test with real content instead of placeholders to replicate authentic usage scenarios.
  - Verify logical and intuitive focus order for VoiceOver and keyboard navigation.
- Tools and Engagement:
  - Utilize Xcode's [Accessibility Inspector](https://developer.apple.com/videos/play/wwdc2019/257/) to identify issues.
  - Involve users with disabilities in beta testing for valuable feedback.
  - Create user-friendly channels for accessibility-related feedback.
- Continuous Improvement:
  - Stay updated with evolving accessibility guidelines and best practices.

>   - Learn about different disabilities to make informed design choices.

Reviewer: [addition] While this is a good call to action, developers might wonder where they can start learning about this information. Consider using the example below to inspire you:
```suggestion
  - Learn about different disabilities to make informed design choices. Sites like [W3C](https://www.w3.org/WAI/tips/) are often recommended to learn about this topic. 
```

```swift
var body: some View {
    Button("Login") {
        // Login action
    }
}
```

This will render a button with default styles and the `"Login"` text, ensuring its visibility to screen readers and adaptive technologies.

## Accessibility Modifiers

SwiftUI provides a range of modifiers to fine-tune the accessibility of an app's elements. Here are some of the most crucial ones:

- `.accessibilityLabel()` and `.accessibilityHint()` modifiers are used to specify custom labels and hints for elements:

    ```swift
    var body: some View {
        Image("info").accessibilityLabel("Information icon")
    }
    ```

  This will render the image while providing a more precise description to assistive technologies.

- `accessibilityValue()` modifier can define the value associated with the element, such as sliders:

    ```swift
    @State private var sliderValue = 50.0
    
    var body: some View {
        Slider(value: $sliderValue, in: 0...100)
            .accessibilityValue("\(Int(sliderValue)) percent")
        Text("Value: \(Int(sliderValue))")
    }
    ```

    This will render a slider while providing context to screen readers about the slider's current value.

> - `.accessibilityTraits()` modifier can indicate the traits of an element:

Author: Some of the traits (like toggle, for example) are beta, but most of them are already available ([see here](https://developer.apple.com/documentation/swiftui/accessibilitytraits)).

    ```swift
    var body: some View {
        TextField("Username", text: $username)
            .focused($username, equals: true)
    }
    ```

    This will render a `TextField`, and when the view appears, the keyboard will automatically be shown, and the cursor will be placed in the `TextField`, ready for user input.

- Using modifiers for Dynamic Type: SwiftUI supports Dynamic Type, which means your app's text can adapt to users' chosen text size:

    ```swift
    var body: some View {
        Text("Accessibility in SwiftUI")
            .font(.body)
            .scaledFont(size: 18)
    }
    ```

    In this example above, the `Text` element is initially set with the [`.font()`](https://www.codecademy.com/resources/docs/swiftui/viewmodifier/font) modifier with `.body` value, and then the `.scaledFont()` modifier is applied to adjust the font size to `18 points`. This ensures the text remains legible even as the user changes their preferred text size settings.

> - Using semantic Views and Tags: Semantic views and tags like `.button`, `.image`, `.heading`, and `.link` for role-specific styling can aid the visual appearance and interactions of the app while also enhancing screen reader users' navigation experiences.

Reviewer 2: list needs to be prefaced
